Implement individual control for kubeadm preflight checks

kad opened the pull request:
**What this PR does / why we need it**:
With new flag `--ignore-preflight-errors` user is able to
decrease severity of each individual check error to warning.

Old flag `--skip-preflight-checks` now acts as `--ignore-preflight-errors=all` and will produce warnings.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubernetes/kubeadm#480

**Special notes for your reviewer**:
/area kubeadm 
/sig cluster-lifecycle

cc @luxas 

**Release note**:
```release-note
New kubeadm flag `--ignore-preflight-errors` that enables to decrease severity of each individual error to warning.
Old flag `--skip-preflight-checks` is marked as deprecated and acts as `--ignore-preflight-errors=all`
```

kad committed:
Implement individual control for kubeadm preflight checks

With new flag `--ignore-checks-errors` user is able to
decrease severity of each individual check to warning.

Old flag `--skip-preflight-checks` now acts as
`--ignore-checks-errors=all` and will produce warnings.

Fixes: kubernetes/kubeadm#480

luxas commented:
/assign @fabriziopandini @xiangpengzhao

anguslees commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -328,3 +329,28 @@ func ValidateAPIEndpoint(c *kubeadm.MasterConfiguration, fldPath *field.Path) fi
+	for _, item := range ignoreChecksErrors {
+		lcItem := strings.ToLower(item) // parameters are case insensitive
+		if ignoreErrors.Has(lcItem) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("ignore-checks-errors"), item, "duplicate value"))
```
Imo this is unnecessarily strict.  If I say --ignore-checks-errors=Foo,Foo then it is unambiguous that I want Foo skipped :P

In particular I can think of many wrapper-automation scenarios where you want to just append values to this field - and with the above check in place that becomes significantly harder (have to check existing values for duplicates first).

kad replied:
That is easy to remove. Original intention was that user must be aware all the time what user tries to ignore. Kind of protection from StackOverflow.

xiangpengzhao replied:
+1 to @anguslees

anguslees commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -328,3 +329,28 @@ func ValidateAPIEndpoint(c *kubeadm.MasterConfiguration, fldPath *field.Path) fi
+	}
+
+	if ignoreErrors.Has("all") && ignoreErrors.Len() > 1 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("ignore-checks-errors"), strings.Join(ignoreErrors.List(), ","), "don't specify individual checks if 'all' is used"))
```
Again, if I say --ignore-checks-errors=Foo,Bar,all then my intention is also unambiguous - I want everything ignored.  The same wrapper-automation use case applies.

kad replied:
Same here: "think twice when copy&paste from stackoverflow". but I'm not insisting, if others will see that as too strict.

xiangpengzhao replied:
+1 to @kad here

kad commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -190,16 +195,21 @@ func AddInitConfigFlags(flagSet *flag.FlagSet, cfg *kubeadmapiext.MasterConfigur
 		skipPreFlight, "skip-preflight-checks", *skipPreFlight,
 		"Skip preflight checks which normally run before modifying the system.",
 	)
+	flagSet.MarkDeprecated("skip-preflight-checks", "it is now equivalent to --ignore-preflight-check-errors=all")
```
Ah. good catch. that's my clean-up error from early draft of this PR. will fix. Thanks @anguslees

kad commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -74,13 +75,14 @@ type Error struct {
 
 func (e *Error) Error() string {
-	return fmt.Sprintf("[preflight] Some fatal errors occurred:\n%s%s", e.Msg, "[preflight] If you know what you are doing, you can skip pre-flight checks with `--skip-preflight-checks`")
+	return fmt.Sprintf("[preflight] Some fatal errors occurred:\n%s%s", e.Msg, "[preflight] If you know what you are doing, you can ignore check's errors with `--ignore-checks-errors`")
```
will rephrase. thx.

xiangpengzhao commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -190,16 +195,21 @@ func AddInitConfigFlags(flagSet *flag.FlagSet, cfg *kubeadmapiext.MasterConfigur
 	)
+	flagSet.StringSliceVar(
+		ignoreChecksErrors, "ignore-checks-errors", *ignoreChecksErrors,
+		"A list of checks whose errors will be shown as warnings. Value 'all' ignores errors from all checks.",
```
It would be good to provide an example (or even a full list of preflight checks). Users may not know what the checks names are.

kad commented:
Updated with addressing review comments:
- duplicate values are not errors now.
- `all` + individual checks is still error
- `all` + `--skip-preflight-checks` is still an error
- rephrased footer message for error output for preflight checks.
- added few examples into help of `--ignore-checks-errors`. List all possible combinations for different invocations is not really maintainable. Check names are written for each warning or error message anyway.
- cleaned typos.

kad commented:
/test pull-kubernetes-node-e2e

anguslees reviewed (approved):
/lgtm

timothysc commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -328,3 +329,27 @@ func ValidateAPIEndpoint(c *kubeadm.MasterConfiguration, fldPath *field.Path) fi
 }
+
+// ValidateIgnoreChecksErrors validates duplicates in ignore-checks-errors flag.
+func ValidateIgnoreChecksErrors(ignoreChecksErrors []string, skipPreflightChecks bool) (sets.String, error) {
```
ValidateIgnoreChecksErrors - doesn't parse well when reading.

timothysc commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -190,16 +195,21 @@ func AddInitConfigFlags(flagSet *flag.FlagSet, cfg *kubeadmapiext.MasterConfigur
 		"Path to kubeadm config file. WARNING: Usage of a configuration file is experimental.",
 	)
+	flagSet.StringSliceVar(
+		ignoreChecksErrors, "ignore-checks-errors", *ignoreChecksErrors,
```
again ignore-checks-errors is difficult to read and rationalize.

timothysc reviewed (changes requested):
The name change on the flag is less intuitive to me.  The primary driver for everything in kubeadm is a UX experience that is easy for folks to understand, but I struggle to understand what "ignore-checks-errors" means.

jbeda commented:
/approve

But I defer to Tim on the issues he has been pointing out.

kad commented:
@timothysc any suggestion on how this flag will be better named ? In early drafts it was named `--ignore-preflight-check-errors`, but then it became more universal (not only pre flight).
I can send one more PR that will update name to something that is more readable.

Regarding warnings of mixed usage: it was in the earlier comments, this is to protect from "copy&paste from StackOverflow" style of usages. If user ignores errors, user needs to understand what and how it is ignored, not randomly mixing values of different flags. `Validate*()` in current implementation are able to return only errors, so warnings are not natural there. However, `--skip-preflight-checks` will produce deprecation warning by other means. See example output in the kubernetes/kubeadm#480

luxas commented:
@kad I think `--ignore-preflight-errors` would be a better name for this, can you make that change please?
Are we all happy with that @timothysc? We should still call this "preflight" though, we just grew these checks to other parts of the codebase

kad commented:
Will do PR soon. do we want to change only flag name or update variable names as well ?

luxas commented:
both, for consistency

timothysc commented:
I'm cool with ```ignore-preflight-errors```

kad commented:
@timothysc it is in PR #56208. I'll update release note here to new name, so release scripts should not pickup wrong name.

luxas commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -328,3 +329,27 @@ func ValidateAPIEndpoint(c *kubeadm.MasterConfiguration, fldPath *field.Path) fi
+	// TODO: remove once deprecated flag --skip-preflight-checks is removed.
+	if skipPreflightChecks {
+		if ignoreErrors.Has("all") {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("ignore-checks-errors"), strings.Join(ignoreErrors.List(), ","), "'all' is used together with deprecated flag --skip-preflight-checks. Remove deprecated flag"))
```
+1. I'd only warn if skipPreflightChecks is there and that's it

luxas commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -910,3 +1055,11 @@ func TryStartKubelet() {
 }
+
+// setHasItemOrAll is helper function that return true if item is present in the set (case insensitive) or special key 'all' is present
+func setHasItemOrAll(s sets.String, item string) bool {
```
unit test for this?
